cmd/segment-reaper/observer.go

```diff
@@ -143,11 +168,139 @@ func (obsvr *observer) processSegment(ctx context.Context, path metainfo.ScopedP
-	// TODO this part will be implemented in next PR
-	// TODO(if): For what is this?
+	for bucket, objects := range obsvr.objects {
+		for path, object := range objects {
+			if object.skip {
+				continue
+			}
+
+			brokenObject := false
+			if !object.hasLastSegment {
+				brokenObject = true
+			} else {
+				includeLastSegment := true
+				segmentsCount := object.segments.Count()
+
+				// using 'expectedNumberOfSegments-1' because 'segments' doesn't contain last segment
+				switch {
+				// this case is only for old style pointers with encrypted number of segments
+				// value 0 means that we don't know how much segments object should have
+				case object.expectedNumberOfSegments == 0:
+					// verify if initial sequence is valid
+					lastSequenceIndex := 0
+					for ; lastSequenceIndex < maxNumOfSegments; lastSequenceIndex++ {
+						has, err := object.segments.Has(lastSequenceIndex)
+						if err != nil {
+							panic(err)
+						}
+						if !has {
+							break
+						}
+					}
+					// unset valid sequence and mark as broken but don't remove last segment
+					for index := 0; index < lastSequenceIndex; index++ {
+						err := object.segments.Unset(index)
+						if err != nil {
+							panic(err)
+						}
+					}
+					// in this case we always mark object as broken because if
+					// there was valid sequence we are removing it from bitmask and leave only
+					// zombie segments
+					brokenObject = true
+					includeLastSegment = false
+				case segmentsCount < int(object.expectedNumberOfSegments)-1:
+					brokenObject = true
+				case segmentsCount > int(object.expectedNumberOfSegments)-1:
+					// verify if initial sequence is valid
+					for index := 0; index < int(object.expectedNumberOfSegments)-1; index++ {
+						has, err := object.segments.Has(index)
+						if err != nil {
+							panic(err)
+						}
+						if !has {
+							brokenObject = true
+							break
+						}
+					}
+					if !brokenObject {
+						// if initial sequence is valid then remove it from object and
+						// mark rest of segments as broken
+						for index := 0; index < int(object.expectedNumberOfSegments)-1; index++ {
+							err := object.segments.Unset(index)
+							if err != nil {
+								panic(err)
+							}
+						}
+						brokenObject = true
+						includeLastSegment = false
+					}
+				case segmentsCount == int(object.expectedNumberOfSegments)-1 && !object.segments.IsSequence():
+					brokenObject = true
+				}
+
+				if brokenObject && includeLastSegment {
+					err := obsvr.printSegment(ctx, "l", bucket, path)
+					if err != nil {
+						return err
+					}
+				}
+			}
+
+			if brokenObject {
+				for index := 0; index < maxNumOfSegments; index++ {
+					has, err := object.segments.Has(index)
+					if err != nil {
+						panic(err)
+					}
+					if has {
+						err := obsvr.printSegment(ctx, "s"+strconv.Itoa(index), bucket, path)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
```

Comment:
Instead of having `brokenObject`, `includeLastSegment` and modifying `object.segments`, can we just have a single `segmentsToDelete []string` and add the segment indices (`l`, `s0`, ..., `sN`) there as they are detected? Then the part for printing the result would just iterate over this `segmentsToDelete` slice.

This may make the code clearer. It should also make it easier to split this method in two smaller: one to detect the zombie segments and another to print them.

Reply:
I agree with what you say about code readability, but if I'm not mistaken, we wanted to reduce the amount of memory needed and reduce allocations.

Reply:
I added a comment on the latest code how to avoid the garbage.